draft-ietf-quic-transport.md

```diff
@@ -3476,10 +3485,12 @@ Datagrams containing Initial packets MAY exceed 1200 bytes if the client
 believes that the Path Maximum Transmission Unit (PMTU) supports the size that
 it chooses.
 
-A server MAY send a CONNECTION_CLOSE frame with error code PROTOCOL_VIOLATION in
-response to an Initial packet it receives from a client if the UDP datagram is
-smaller than 1200 bytes. It MUST NOT send any other frame type in response, or
-otherwise behave as if any part of the offending packet was processed as valid.
+A server that has no existing state for a connection MUST discard an Initial
+packet that is carried in a UDP datagram that is smaller than 1200 bytes.  Other
+packets in the datagram SHOULD also be discarded.  A server MAY send a
+CONNECTION_CLOSE frame with error code PROTOCOL_VIOLATION in addition to
+discarding a packet if that does not affect a connection for which the server
+has established state; see {{immediate-close}}.
```

Comment:
I am confused by the last sentence. Is it saying that when a server receives a small Initial packet for a connection for which it already has state, the server MAY send a PROTOCOL_VIOLATION, but it must retain the state for that connection? Assuming that is the case, I wonder if there is a value in having such a requirement. I am not sure if there is a case such a behavior helps us mitigate attacks.

Reply:
My understanding was essentially that it's OK for the server to send PROTOCOL_VIOLATION in response to an Initial for which there isn't connection state already (i.e. the first Initial packet for a connection); because that means it couldn't affect a pre-existing connection.

Reply:
Yeah if we interpret the last sentence that way, I think it would contradict with the "MUST discard" in the first sentence?

Reply:
I'm confused here too. What does it mean to not affect the connection if the server just sent out a CONNECTION_CLOSE for it?

Reply:
I am in agreement with @ianswett : perhaps it's best to recommend dropping the connection and not send anything in response. This is a simple enough error for a client to avoid.

Reply:
The problem I'm trying to finesse around here is one where the server gets a partial ClientHello in an Initial.  That is processed and held.  The next packet is a small Initial.  The server should discard that *without* generating a CONNECTION_CLOSE or it will disrupt the in-progress connection establishment.

Reply:
My understanding is that the server should discard that packet because it's not 1200 bytes and it has no state for the connection?

Reply:
@martinthomson Thank you for the clarification.

I think there are three possible behaviors in such a situation:
* immediate-close the connection with PROTOCOL_VIOLATION
* discard the small Initial
* discard the small Initial, send PROTOCOL_VIOLATION, without affecting the connection state

I think that either of the first two makes sense.

However, the third one (which IIUC is what the text suggests) does not. This is because sending PROTOCOL_VIOLATION will instruct the client to discard *its* connection state, regardless of the sender of the small Initial packet being the client or an attacker. There is no point in the server retaining the connection state after sending PROTOCOL_VIOLATION.

Therefore, I think that the advice should be: SHOULD discard, MAY immediate-close.

Reply:
@kazuho, I agree with you about the first two.  The rule I'm proposing is MUST discard, but MAY immediate-close in addition IF that won't affect state the server holds.  Yes, that might affect state the client holds (but the server does not), but that's a risk the server takes.  It's a trade-off about the level of feedback the server wants to provide vs. the risk of DoS that the server has to make.

Are we agreeing violently, but talking about how to phrase the recommendation to not immediate-close?

Reply:
@martinthomson 
> The rule I'm proposing is MUST discard, but MAY immediate-close in addition IF that won't affect state the server holds.

I think we might have confusion in what "immediate-close" means? My understanding is that "immediate-close" is an action that affects state. It is defined as entering the closing state, then draining state. If the intention is to suggest that a server might send PROTOCOL_VIOLATION without affecting the connection state, then it is confusing to use the term "immediate-close."

That said, what I stated in [my comment right above](https://github.com/quicwg/base-drafts/pull/3292#discussion_r355760171) is that I think a state-*ful* "immediate-close" is fine, but sending PROTOCOL_VIOLATION without affecting connection state does not make sense.

Reply:
We've now unfortunately lost this comment in the diffs :-/ But that was my confusion too -- sending CONNECTION_CLOSE changes state. I don't understand what it means to send PROTOCOL_VIOLATION without affecting state.

Reply:
Let's see if I can keep this thread alive.  This is sending a packet with a CONNECTION_CLOSE, not an immediate close (I was under the impression that the shorthand was OK here, but it was clearly a mistake).  The goal is to use CONNECTION_CLOSE as a signal only, with the intent of triggering the corresponding logic at the client.  But the server is discarding this packet and so it won't have any connection state.  This won't result in creation or destruction of server state.

Reply:
> The goal is to use CONNECTION_CLOSE as a signal only, with the intent of triggering the corresponding logic at the client.

I think it is not a bad idea to have a MAY for that.

> But the server is discarding this packet and so it won't have any connection state. This won't result in creation or destruction of server state.

I do not think that the sentences reflect what you think. Quoted below are the last two sentences of this paragraph. It is my understanding that the intention of the last sentence is to clarify the rationale behind the previous sentence, but as you can see from the emphasized text, they are discussing about different conditions.

> _A server MAY send a CONNECTION_CLOSE frame with error code PROTOCOL_VIOLATION in addition to discarding a packet if that does not affect **a connection for which the server has previously established state**.  Sending CONNECTION_CLOSE will not affect server state in the same way as an immediate close ({{immediate-close}}) **as the server has no state**, but it will cause any client to terminate a connection attempt._

I think what we actually want to state is something like:  _A server that has no existing state for a connection MUST discard an Initial packet that is carried in a UDP datagram that is smaller than 1200 bytes.  Other packets in that datagram SHOULD also be discarded. In addition to discarding the packet, a server MAY respond with a CONNECTION_CLOSE frame with error code PROTOCOL_VIOLATION, without creating state. When a server receives such an Initial packet for which it already has connection state, it MUST either drop that packet or close the connection with error code PROTOCOL_VIOLATION ({{immediate-close}})._